apis/goharbor.io/v1alpha2/harborcluster_webhook.go

```diff
@@ -0,0 +1,164 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// Log used this webhook
+var clog = logf.Log.WithName("harborcluster-resource")
+
+// SetupWebhookWithManager sets up validating webhook of HarborCluster
+func (hc *HarborCluster) SetupWebhookWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(hc).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-goharbor-io-v1alpha2-harborcluster,mutating=false,failurePolicy=fail,groups=goharbor.io,resources=harborclusters,versions=v1alpha2,name=vharborcluster.kb.io
+
+var _ webhook.Validator = &HarborCluster{}
+
+func (hc *HarborCluster) ValidateCreate() error {
+	clog.Info("validate creation", "name", hc.Name, "namespace", hc.Namespace)
+
+	return hc.validate()
+}
+
+func (hc *HarborCluster) ValidateUpdate(old runtime.Object) error {
+	clog.Info("validate updating", "name", hc.Name, "namespace", hc.Namespace)
+
+	return hc.validate()
+}
+
+func (hc *HarborCluster) ValidateDelete() error {
+	clog.Info("validate deletion", "name", hc.Name, "namespace", hc.Namespace)
+
+	return nil
+}
+
+func (hc *HarborCluster) validate() error {
+	var allErrs field.ErrorList
+
+	// For database(psql), cache(Redis) and storage, either external services or in-cluster services MUST be configured
+	if err := hc.validateStorage(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := hc.validateDatabase(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := hc.validateCache(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(schema.GroupKind{Group: GroupVersion.Group, Kind: "HarborCluster"}, hc.Name, allErrs)
+}
+
+func (hc *HarborCluster) validateStorage() *field.Error {
+	// Storage
+	// External is not configured
+	if err := hc.Spec.ImageChartStorage.Validate(); err != nil {
+		clog.Info("validate spec.imageChartStorage", "cause", err.Error())
+
+		// And in-cluster minIO is not configured
+		if hc.Spec.InClusterStorage == nil {
+			// Invalid and not acceptable
+			return field.Invalid(
+				field.NewPath("spec").
+					Child("imageChartStorage", "inClusterStorage"),
+				hc.Spec.ImageChartStorage,
+				"both storage and in-cluster storage are not correctly configured",
+			)
+		}
+	} else {
+		if hc.Spec.InClusterStorage != nil {
+			// Both are configured, conflict
+			return field.Invalid(
+				field.NewPath("spec").
+					Child("imageChartStorage", "inClusterStorage"),
+				hc.Spec.InClusterStorage,
+				"conflicts: both storage and in-cluster storage are configured, only one is required to set",
+			)
+		}
+	}
+
+	return nil
+}
+
+func (hc *HarborCluster) validateDatabase() *field.Error {
```

Comment:
Do we need to validate in cluster database？

Reply:
we'll validate all, only one of them configured is valid.

no one is set all both are set are all invalid cases